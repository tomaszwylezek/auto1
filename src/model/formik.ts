import React from 'react';

export interface FormikField {
  onChange: (e: React.ChangeEvent<any>) => void;
  onBlur: (e: any) => void;
  value: any;
  name: string;
}
