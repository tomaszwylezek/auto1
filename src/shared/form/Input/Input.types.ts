export interface IInputOwnProps {
  label?: string;
  type?: string;
  className?: string;
  placeholder?: string;
}

export interface IInputProps extends IInputOwnProps {}
