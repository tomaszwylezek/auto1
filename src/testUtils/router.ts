import { RouteComponentProps } from 'react-router';
import { UnregisterCallback, Href, Location } from 'history';

export function getMockedRouterProps<P = any>(params?: P, locationData?: Partial<Location>) {
  const location: Location = {
    pathname: '',
    hash: '',
    key: '',
    search: '',
    state: '',
    ...locationData
  };

  const props: RouteComponentProps<P> = {
    match: {
      isExact: true,
      params,
      path: '',
      url: ''
    },
    location,
    history: {
      length: 2,
      action: 'POP',
      location,
      push: jest.fn(),
      replace: () => null,
      go: () => null,
      goBack: jest.fn(),
      goForward: () => null,
      block: () => {
        return null as UnregisterCallback;
      },
      createHref: () => {
        const temp: Href = '';
        return temp as Href;
      },
      listen: () => {
        return null as UnregisterCallback;
      }
    },
    staticContext: {}
  };

  return props;
}
