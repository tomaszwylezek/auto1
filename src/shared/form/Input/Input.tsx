import React from 'react';
import { get } from 'lodash';
import { FieldProps } from 'formik';
import classNames from 'classnames';

import { Label } from '../Label/Label';

// Models
import { IInputProps } from './Input.types';

export const Input: React.FC<IInputProps & FieldProps> = ({
  field,
  form: { touched, setFieldTouched },
  label,
  type = 'text',
  required = false,
  className,
  placeholder,
  ...props
}): JSX.Element => {
  const isTouched: boolean = !!get(touched, field.name);
  const inputClassName: string = classNames('form-control', className);

  const input: JSX.Element = (
    <>
      <input
        {...field}
        {...props}
        placeholder={placeholder ? placeholder : undefined}
        className={inputClassName}
        type={type}
        onChange={(e: React.FormEvent<HTMLInputElement>): void => {
          if (!isTouched) {
            setFieldTouched(field.name, true, true);
          }
          field.onChange(e);
        }}
        id={field.name}
      />
    </>
  );

  return (
    <div className="form-group">
      {!!label && (
        <Label id={field.name}>
          {label}
          {required && <span className="text-danger"> *</span>}
        </Label>
      )}
      {input}
    </div>
  );
};
