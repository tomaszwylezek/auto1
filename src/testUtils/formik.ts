import { FormikActions, FormikHandlers, FormikProps } from 'formik';
import { FormikField } from 'model';

export function getMockedFormikConfig<T>(options: Partial<FormikProps<T>> = {}): FormikProps<T> {
  return {
    ...getMockedFormikActions(),
    ...getMockedFormikHandles(),
    enableReinitialize: false,
    isValidating: false,
    isSubmitting: false,
    dirty: false,
    isValid: false,
    submitCount: 0,
    registerField: jest.fn(),
    unregisterField: jest.fn(),
    values: {} as T,
    errors: {} as T,
    initialValues: {} as T,
    touched: {} as T,
    ...options
  };
}

export function getMockedFormikActions<V>(options: Partial<FormikActions<V>> = {}): FormikActions<V> {
  return {
    setStatus: jest.fn(),
    setError: jest.fn(),
    setErrors: jest.fn(),
    setSubmitting: jest.fn(),
    setTouched: jest.fn(),
    setValues: jest.fn(),
    setFieldValue: jest.fn(),
    setFieldError: jest.fn(),
    setFieldTouched: jest.fn(),
    validateForm: jest.fn(),
    validateField: jest.fn(),
    resetForm: jest.fn(),
    submitForm: jest.fn(),
    setFormikState: jest.fn(),
    ...options
  };
}

export function getMockedFormikHandles(options: Partial<FormikHandlers> = {}): FormikHandlers {
  return {
    handleSubmit: jest.fn(),
    handleReset: jest.fn(),
    handleBlur: jest.fn(),
    handleChange: jest.fn(),
    ...options
  };
}

export function getMockedField(options: Partial<FormikField> = {}): FormikField {
  return {
    onChange: jest.fn(),
    onBlur: jest.fn(),
    value: '',
    name: '',
    ...options
  };
}
