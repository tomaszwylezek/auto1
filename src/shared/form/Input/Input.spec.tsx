import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { Input } from './Input';

// Config
import { getMockedField, getMockedFormikConfig } from 'testUtils/formik';

// Models
import { FieldProps } from 'formik';
import { IInputProps } from './Input.types';

describe('Input component should', (): void => {
  let component: ShallowWrapper<IInputProps & FieldProps>;
  let props: IInputProps & FieldProps;

  beforeEach((): void => {
    props = {
      label: 'bar',
      field: getMockedField({ name: 'foo' }),
      form: getMockedFormikConfig()
    };

    component = shallow(<Input {...props} />);
  });

  it('renders correctly', (): void => {
    expect(component).toMatchSnapshot();
  });
});
