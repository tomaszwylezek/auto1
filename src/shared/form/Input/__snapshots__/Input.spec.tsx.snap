// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should renders correctly 1`] = `
<div
  className="form-group"
>
  <Label
    id="foo"
  >
    bar
  </Label>
  <input
    className="form-control"
    id="foo"
    name="foo"
    onBlur={[MockFunction]}
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;
